### PR TITLE
Refactor parallel dom traversal to be agnostic to the processing steps themselves

### PR DESCRIPTION
**Pull request by bholley**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9103)

### CODE REVIEW

**Comment by bholley** [comment]
r? @SimonSapin 

CC @Manishearth, who might be interested in the solution I settled on (which is more design-y and less type-system-hackery).

File: components/layout/parallel.rs [hunk 1]
```diff
@@ -441,12 +384,14 @@ fn run_queue_with_custom_work_data_type<To, F>(
...
+pub fn traverse_dom_preorder<'ln, N, T>(
+                             root: N,
                              shared_layout_context: &SharedLayoutContext,
-                             queue: &mut WorkQueue<SharedLayoutContext, WorkQueueData>) {
+                             queue: &mut WorkQueue<SharedLayoutContext, WorkQueueData>)
+                             where N: LayoutNode<'ln>, T: DomTraversal<'ln, N> {
     run_queue_with_custom_work_data_type(queue, |queue| {
         queue.push(WorkUnit {
-            fun:  recalc_style::<'ln, ConcreteLayoutNode>,
+            fun:  top_down_dom::<'ln, N, T>,
```
Comment (Manishearth) [inline components/layout/parallel.rs]: no need for the lifetime here

**Pull request closed without merging**